zilliqa/src/exec.rs

```diff
@@ -673,23 +676,101 @@ impl State {
         value: u128,
     ) -> Result<u64> {
         let gas_price = gas_price.unwrap_or(GAS_PRICE);
-        let gas = gas.unwrap_or(BLOCK_GAS_LIMIT);
 
+        let mut right = gas.unwrap_or(MAX_EVM_GAS_LIMIT).0;
+        let mut left = lower_bound_gas_estimate(to_addr, &data);
+
+        if gas.is_some() {
+            right = gas.unwrap().0;
+        }
+
+        let upper_bound = right;
+
+        // Check if estimation succeeds with the highest possible gas
+        self.estimate_gas_inner(
+            from_addr,
+            to_addr,
+            data.clone(),
+            chain_id,
+            current_block,
+            EvmGas(upper_bound),
+            gas_price,
+            value,
+        )?;
+
+        // We don't do deeper search if the range is smaller than the following param
```

Comment:
wording nitpick, sorry
```suggestion
        // We don't do deeper search if the difference between `min` and `max` is smaller than the following param
```